### PR TITLE
Add value and constructor path

### PR DESCRIPTION
**Pull request by panglesd**
The ability to resolve those path will be useful in many render source code new features. In particular, this is a prerequisite to #976.

### CODE REVIEW

File: src/xref2/link.ml [hunk 1]
```diff
@@ -143,6 +153,52 @@ let type_path : Env.t -> Paths.Path.Type.t -> Paths.Path.Type.t =
...
+        match Tools.resolve_constructor_path env cp with
+        | Ok p' ->
+            let result = Tools.reresolve_constructor env p' in
+            `Resolved Lang_of.(Path.resolved_constructor (empty ()) result)
+        | Error e ->
+            Errors.report ~what:(`Constructor_path cp) ~tools_error:e `Lookup;
+            let _ =
+              ignore e;
+              failwith "todo"
+            in
```
Comment (gpetiot) [inline src/xref2/link.ml:194]: To clean up.

File: src/xref2/tools.ml [hunk 2]
```diff
@@ -1110,6 +1252,148 @@ and resolve_type :
...
+      | `FType_removed _ -> (* failwith "error" *) Error `Find_failure
+      | `FType (_, t) ->
+          handle_constructor_lookup env (ConstructorName.to_string id) parent t
+      (* let result = *)
+      (*   match Find.value_in_sig parent_sig (ValueName.to_string id) with *)
+      (*   | `FValue (name, t) :: _ -> *)
+      (*       Some (`Value (parent, name), `FValue (name, Subst.value sub t)) *)
+      (*   | [] -> None *)
+      (* in *)
+      (* of_option ~error:`Find_failure result *))
```
Comment (gpetiot) [inline src/xref2/tools.ml:1391]: To clean up.

File: src/xref2/tools.ml [hunk 3]
```diff
@@ -886,6 +972,62 @@ and lookup_type :
...
+  |> map_error (fun e -> (`ParentC e :> simple_constructor_lookup_error))
+  >>= fun t ->
+  match t with
+  | `FType (_, t) ->
+      handle_constructor_lookup env (ConstructorName.to_string name) parent t
+      >>= fun (_, x) -> Ok x
+  | `FType_removed _ -> Error `Find_failure (* failwith "Error" *)
+
+(* handle_value_lookup env (ValueName.to_string id) p sg *)
+(* >>= fun (_, `FValue (name, c)) -> Ok (`FValue (name, Subst.value sub c)) *)
```
Comment (gpetiot) [inline src/xref2/tools.ml:1029]: To clean up.

File: src/xref2/compile.ml [hunk 4]
```diff
@@ -53,6 +74,11 @@ and class_type_path : Env.t -> Paths.Path.ClassType.t -> Paths.Path.ClassType.t
       | Ok p' -> `Resolved Lang_of.(Path.resolved_class_type (empty ()) p')
       | Error _ -> p)
 
+let () =
+  (* Until those are used *)
+  ignore value_path;
+  ignore constructor_path
```
Comment (gpetiot) [inline src/xref2/compile.ml:80]: Would it break things to modify `source_info_infos` like in #976, or would it allow to check more things in the tests?

File: src/xref2/link.ml [hunk 5]
```diff
@@ -143,6 +153,52 @@ let type_path : Env.t -> Paths.Path.Type.t -> Paths.Path.Type.t =
...
+            p)
+
+let () =
+  (* Until those are used *)
+  ignore value_path;
+  ignore constructor_path
```
Comment (gpetiot) [inline src/xref2/link.ml:200]: Same question here, would it break things to update the `source_info` in `unit` like in #976?

**Review by gpetiot — APPROVED** [review]
I left a few questions (and some things should be cleaned up) but otherwise it looks good to me!

File: src/xref2/link.ml (hunk 1 above)
Comment (panglesd) [reply src/xref2/link.ml:194]: Woops, slipped into my staging area... thanks!

File: src/xref2/compile.ml (hunk 4 above)
Comment (panglesd) [reply src/xref2/compile.ml:80]: I'm not sure about the question! Initially, #976 was a single PR, but I turned it into three PRs:
- one for the path resolution (this one)
- one for the refactoring #1031
- #976 which will be rebased on top of master when the two other one have been merged.

So no, it won't break things to have `source_infos` as in #976, on the contrary that is the goal!

Comment (gpetiot) [reply src/xref2/compile.ml:80]: Go it, thanks!

**Comment by jonludlam** [comment]
I'm not very convinced with all the datatype stuff; I think we've not got the best set of types here. Having said that, we're already in this situation and this PR won't make it more tricky to rip it out, and there are other PRs that depend upon this. Hence, let's figure this out later.

**Pull request merged**